File: cli/python/src/memgo_cli/cli/context.py
"""命令依赖装配、鉴权预检和遥测触发。"""

from __future__ import annotations

from dataclasses import replace

import typer
from rich.console import Console

from memgo_cli.backend.json import JsonObject
from memgo_cli.backend.types import Backend, BackendContext
from memgo_cli.config.models import MemGoConfig
from memgo_cli.output.branding import print_error
from memgo_cli.runtime.state import (
    caller_type,
    capture_notice,
    register_cleanup,
    set_validated_email,
    validated_email,
)
from memgo_cli.runtime.warning import warn_optional

console = Console()
err_console = Console(stderr=True)


def _fire_telemetry(command_name: str, extra: JsonObject | None) -> None:
    """异步发送命令遥测，可选集成失败时记录告警。"""
    try:
        from memgo_cli.integrations.telemetry import capture_event

        props: JsonObject = {"command": command_name}
        if extra:
            props.update(extra)
        capture_event(f"cli.{command_name}", props, pre_resolved_email=validated_email())
    except (OSError, ValueError) as error:
        warn_optional("telemetry", error)


def _get_backend_and_config(
    api_key: str | None,
    base_url: str | None,
) -> tuple[Backend, MemGoConfig]:
    """加载配置并创建连接器，预检凭据且缓存平台邮箱。"""
    from memgo_cli.backend import get_backend
    from memgo_cli.backend.errors import AuthError
    from memgo_cli.config.store import load_config, save_config

    config = load_config()

    if api_key:
        config = replace(config, platform=replace(config.platform, api_key=api_key))
    if base_url:
        config = replace(config, platform=replace(config.platform, base_url=base_url))

    if not config.platform.api_key:
        print_error(
            err_console,
            "No API key configured.",
            hint="Run 'memgo init' or set MEMGO_API_KEY environment variable.",
        )
        raise typer.Exit(1)

    backend = get_backend(config, BackendContext(caller_type, capture_notice))
    register_cleanup(backend.close)

    # 使用短超时预检密钥
    try:
        ping_data = backend.ping(timeout=5.0)
        email = ping_data.get("user_email") if isinstance(ping_data, dict) else None
        if isinstance(email, str) and email:
            set_validated_email(email)
            if config.platform.user_email != email:
                config = replace(config, platform=replace(config.platform, user_email=email))
                save_config(config)
    except AuthError:
        print_error(
            err_console,
            "Invalid or expired API key.",
            hint="Run 'memgo init' or set MEMGO_API_KEY environment variable.",
        )
        raise typer.Exit(1) from None

    return backend, config


def _get_backend(
    api_key: str | None,
    base_url: str | None,
) -> Backend:
    """创建已完成凭据预检的 Platform 连接器。"""
    backend, _config = _get_backend_and_config(api_key, base_url)
    return backend

File: cli/node/src/cli/context.ts
import { AuthError, type Backend, getBackend } from '../backend/index.js';
import { loadConfig, saveConfig } from '../config/store.js';
import type { MemGoConfig } from '../config/types.js';
import { printError } from '../output/branding.js';
import { exit } from '../runtime/exit.js';
import { captureNotice, isAgentMode } from '../runtime/state.js';

/**
 * 合并调用参数与配置，验证连接并更新账号邮箱。
 * @param apiKey - 访问服务的 API 密钥。
 * @param baseUrl - 服务基础地址。
 * @returns 通过预检的后端与有效配置。
 */
export async function getBackendAndConfig(
  apiKey?: string,
  baseUrl?: string
): Promise<{ backend: Backend; config: MemGoConfig }> {
  const saved = loadConfig();
  const config = {
    ...saved,
    platform: {
      ...saved.platform,
      apiKey: apiKey || saved.platform.apiKey,
      baseUrl: baseUrl || saved.platform.baseUrl,
    },
  };

  if (!config.platform.apiKey) {
    printError(
      'No API key configured.',
      "Run 'memgo init' or set MEMGO_API_KEY environment variable."
    );
    exit(1);
  }

  const backend = getBackend(config, {
    callerType: () => (isAgentMode() ? 'agent' : 'user'),
    notice: captureNotice,
  });

  try {
    const pingData = await backend.ping();

    const email = pingData?.user_email as string | undefined;
    if (email) {
      if (config.platform.userEmail !== email) {
        config.platform.userEmail = email;
        saveConfig(config);
      }
    }
  } catch (e) {
    if (e instanceof AuthError) {
      printError(
        'Invalid or expired API key.',
        "Run 'memgo init' or set MEMGO_API_KEY environment variable."
      );
      exit(1);
    }
    throw e;
  }

  return { backend, config };
}

/**
 * 完成配置与鉴权预检后取得后端连接器。
 * @param apiKey - 访问服务的 API 密钥。
 * @param baseUrl - 服务基础地址。
 * @returns 通过预检的后端连接器。
 */
export async function getBackendOnly(apiKey?: string, baseUrl?: string): Promise<Backend> {
  return (await getBackendAndConfig(apiKey, baseUrl)).backend;
}
